### PR TITLE
Implement Event class for creating logs. 

### PR DESCRIPTION
**Pull request by miladHakimi**
Currently, the logging mechanism is unstructured. Various log functions with different interfaces are used to generate the logs. All these functions receive a series of inputs, combine them in a string with csv format, and write them in the standard output or a file.
`The`Event` class unifies the interface for creating logs. It provides these two functionalities:
1) Using the same interface for creating all the logs to replace the current ad-hoc event logging functions.
2) Allowing log generation for different formats other than csv.

Each `Event` has a name and a set of attributes. Attributes contain the value and its name. The name is a string telling what the attribute is (e.g, duration, timestamp, etc).

The current log functions use only a limited set of types; string, int64, and vector. Hence, right now, `Attribute` has three implementations covering the mentioned types.

This PR also contains the implementation for Event, an event-to-csv generator, and simple unit tests to check if they work as expected.

Issue: #75

### CODE REVIEW

File: units/event_log_tests.cc [hunk 1]
```diff
@@ -0,0 +1,132 @@
+#include <array>
```
Comment (kuhar) [inline units/event_log_tests.cc]: Could you remove this and any other unused includes?

File: units/event_log_tests.cc [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <cstdio>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include "gtest/gtest.h"
+
```
Comment (kuhar) [inline units/event_log_tests.cc:8]: Could you put all the code below in an anonymous namespace? This way we won't have to worry about any name clashes and [ODR violations](https://en.wikipedia.org/wiki/One_Definition_Rule).

File: units/event_log_tests.cc [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+// TODO(miladhakimi): Need better naming.
+enum LogLevel { CommonFile, PrivateFile };
+enum ValueType { String, Int64, Vector };
```
Comment (kuhar) [inline units/event_log_tests.cc]: Can `Vector` be used to compose other `ValueType`s?  If Vector can only hold one type, should this be `VectorInt64` instead?

File: units/event_log_tests.cc [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+#include <cstdint>
+#include <cstdio>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include "gtest/gtest.h"
+
+// TODO(miladhakimi): Need better naming.
+enum LogLevel { CommonFile, PrivateFile };
```
Comment (kuhar) [inline units/event_log_tests.cc]: Could you add a short comment for this and other types/functions you plan to expose in the final public interface?

File: units/event_log_tests.cc [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+std::string ValueToString(const std::string &value) { return value; }
+
+std::string ValueToString(const int64_t &value) {
+  return std::to_string(value);
+}
+
+// TODO(miladhakimi): This can be different for other formats.
+template <typename T>
+std::string ValueToString(const T &values) {
```
Comment (kuhar) [inline units/event_log_tests.cc]: I think it would be better to introduce an overload for `vector<int64_t>` only. This will simplify the code and won't make others think that more types are supported

File: units/event_log_tests.cc [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+#include "gtest/gtest.h"
+
+// TODO(miladhakimi): Need better naming.
+enum LogLevel { CommonFile, PrivateFile };
+enum ValueType { String, Int64, Vector };
+
+std::string ValueToString(const std::string &value) { return value; }
+
+std::string ValueToString(const int64_t &value) {
```
Comment (kuhar) [inline units/event_log_tests.cc]: Ints are cheap to copy, pass it by value

File: units/event_log_tests.cc [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+  return std::to_string(value);
+}
+
+// TODO(miladhakimi): This can be different for other formats.
+template <typename T>
+std::string ValueToString(const T &values) {
+  std::string tmp = "\"[";
```
Comment (kuhar) [inline units/event_log_tests.cc]: `std::stringstream` should be more efficient (in terms of the number of allocations) for incrementally building long strings.

File: units/event_log_tests.cc [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+  return tmp;
+}
+
+struct Attribute {
+  const char *name;
+  ValueType value_type;
+
+  template <typename T>
+  T *cast() {
+    return static_cast<T *>(this);
```
Comment (kuhar) [inline units/event_log_tests.cc]: This looks unsafe, we don't check if the cast is possible/legal. We should either check that the value type matches the requested type `T` or drop this function entirely and perform `static_cast`s directly

File: units/event_log_tests.cc [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+  T &GetValue() { return value; }
+};
+
+using VectorAttr = AttributeImpl<std::vector<int64_t>, ValueType::Vector>;
+using StringAttr = AttributeImpl<std::string, ValueType::String>;
+using Int64Attr = AttributeImpl<int64_t, ValueType::Int64>;
+
+struct EventLog {
+  const char *name;
+  std::vector<Attribute *> attributes;
```
Comment (kuhar) [inline units/event_log_tests.cc]: I think this class matches the 'LogabbleEvent` class in the original issue (https://github.com/googlestadia/performance-layers/issues/69), but the name suggests it's the log of events. This seems misleading to me.

File: units/event_log_tests.cc [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+  LogLevel log_level;
+
+  EventLog(const char *_name, const std::vector<Attribute *> &_attributes,
+           LogLevel _log_level)
+      : name(_name), attributes(_attributes), log_level(_log_level) {}
+
+  const char *GetEventName() { return name; }
+  const std::vector<Attribute *> &GetAttributes() { return attributes; }
```
Comment (kuhar) [inline units/event_log_tests.cc]: What do you think about making this function `virtual` and letting the derived classes override it? The reason is that when some class derives from `EventLog` it must first run an `EventLog`'s constructor. If attributes are members in the derived class, you may not get the change to properly initialize them before you have to call the base constructor with attribute pointers.

File: units/event_log_tests.cc [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+std::string EventToCSVString(EventLog &event) {
+  const std::vector<Attribute *> &attributes = event.GetAttributes();
+
+  std::string csv_str = event.GetEventName();
+  csv_str += ",";
+  size_t e;
+  for (size_t i = 0, e = attributes.size(); i != e; ++i) {
+    switch (attributes[i]->value_type) {
+      case ValueType::Int64: {
+        csv_str += ValueToString(attributes[i]->cast<Int64Attr>()->GetValue());
```
Comment (kuhar) [inline units/event_log_tests.cc]: This `ValueToString` implementation/format seems specific to CSV, especially when we consider array printing. What do you think about calling it `ValueToCSVString` instead?

File: units/event_log_tests.cc [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+      }
+      case ValueType::String: {
+        csv_str += ValueToString(attributes[i]->cast<StringAttr>()->GetValue());
+        break;
+      }
+      case ValueType::Vector: {
+        csv_str += ValueToString(attributes[i]->cast<VectorAttr>()->GetValue());
+        break;
+      }
+      default:
```
Comment (kuhar) [inline units/event_log_tests.cc]: Some compilers warn about a default switch cases when the other cases are already exhaustive. I think we should remove this default case and instead rely on compiler warnings.

File: units/event_log_tests.cc [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+    if (i + 1 != e) csv_str += ",";
+  }
+  return csv_str;
+}
+
+TEST(EventLog, AttributeCreation) {
+  Int64Attr compile_timestamp("timestamp", 1601314733266018318);
+  StringAttr state("state", "1");
+  VectorAttr pipeline("pipeline", {0x67d6fd0aaa78a6d8, 0x67d390249c2f20ce});
+  SUCCEED();
```
Comment (kuhar) [inline units/event_log_tests.cc]: This doesn't seem to check anything. Maybe we could check that the number of attributes is as expected and that the attribute pointers match the local variables?

**Review by kuhar — CHANGES_REQUESTED** [review]
> The current log functions use only a limited set of types; string, int64, and vector. Hence, right now, Attribute has three implementations covering the mentioned types.

> Fixes: https://github.com/googlestadia/performance-layers/issues/75

This PR doesn't fully fix the issue, we should use `Issue: <link>` instead

File: units/event_log_tests.cc (hunk 5 above)
Comment (miladHakimi) [reply units/event_log_tests.cc]: Makes sense.

File: units/event_log_tests.cc (hunk 11 above)
Comment (miladHakimi) [reply units/event_log_tests.cc]: Yeah. True. I put a comment on top to fix it. The name is general but the functionality is specific to csv string.

File: CMakeLists.txt [hunk 14]
```diff
@@ -135,6 +135,7 @@ enable_testing()
 add_executable(layer_support_tests
     units/input_buffer_tests.cc
     units/log_scanner_tests.cc
+    units/event_log_tests.cc
```
Comment (kuhar) [inline CMakeLists.txt]: Could you put this unit test file first? It's supposed to be in the alphabetical order.

File: units/event_log_tests.cc [hunk 15]
```diff
@@ -0,0 +1,183 @@
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <cstdio>
+#include <iostream>
+#include <ostream>
```
Comment (kuhar) [inline units/event_log_tests.cc]: I don't see these three includes used, can we drop them?

File: units/event_log_tests.cc [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+std::string ValueToCSVString(const int64_t value) {
+  return std::to_string(value);
+}
+
+std::string ValueToCSVString(const std::vector<int64_t> &values) {
+  std::string tmp = "\"[";
+  using std::begin;
+  using std::end;
+  auto e = end(values);
+  for (auto it = begin(values); it != e; ++it) {
```
Comment (kuhar) [inline units/event_log_tests.cc]: This loop no longer needs to be generic, we should change this to a plain for loop over all indices. Something like:
```c++
for (size_t i  = 0, e = values.end(); i != e; ++i)
```

File: units/event_log_tests.cc [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+  tmp += "]\"";
+  return tmp;
+}
+
+struct Attribute {
+  const char *name;
+  AttributeType value_type;
+
+  template <typename T>
+  bool isa() {
```
Comment (kuhar) [inline units/event_log_tests.cc]: Could you add a comment of explanation for this function?

File: units/event_log_tests.cc [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+  for (auto it = begin(values); it != e; ++it) {
+    const char *delimiter = std::next(it) != values.end() ? "," : "";
+    tmp += std::to_string(*it) + delimiter;
+  }
+  tmp += "]\"";
+  return tmp;
+}
+
+struct Attribute {
+  const char *name;
```
Comment (kuhar) [inline units/event_log_tests.cc]: I think it would be better for `Attribute` to be a `class` with clearly defined public and private interface. For example, we probably don't want all users of `Attribute` to be able to change the name, which we can enforce by making the field `private` and exposing a getter `getName() const;`.

File: units/event_log_tests.cc [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+  AttributeType value_type;
+
+  template <typename T>
+  bool isa() {
+    return value_type == T::id;
```
Comment (kuhar) [inline units/event_log_tests.cc]: Because this code requires all derived classes to expose `id`, I think we should document this in a comment above `Attribute`

File: units/event_log_tests.cc [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+  auto e = end(values);
+  for (auto it = begin(values); it != e; ++it) {
+    const char *delimiter = std::next(it) != values.end() ? "," : "";
+    tmp += std::to_string(*it) + delimiter;
+  }
+  tmp += "]\"";
+  return tmp;
+}
+
+struct Attribute {
```
Comment (kuhar) [inline units/event_log_tests.cc]: Documentation is missing

File: units/event_log_tests.cc [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+// Valid typenames: {std::string, int64_t, std::vector<int64_t>}
+// Example:
+//    AttributeImpl<int64_t, AttributeType::Int64> timestamp("timestamp", 1234);
+template <typename T, AttributeType VT>
+struct AttributeImpl : Attribute {
+  T value;
+  static constexpr AttributeType id = VT;
+
+  AttributeImpl(const char *_name, T _value)
+      : Attribute{_name, VT}, value(_value) {}
```
Comment (kuhar) [inline units/event_log_tests.cc]: nit: the names don't conform to the style guide: https://google.github.io/styleguide/cppguide.html#Naming
This should be `(const char *name, T value)`

File: units/event_log_tests.cc [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+// Implements the Attribute base class for each specific `AttributeType`.
+// Valid typenames: {std::string, int64_t, std::vector<int64_t>}
+// Example:
+//    AttributeImpl<int64_t, AttributeType::Int64> timestamp("timestamp", 1234);
+template <typename T, AttributeType VT>
+struct AttributeImpl : Attribute {
+  T value;
+  static constexpr AttributeType id = VT;
+
+  AttributeImpl(const char *_name, T _value)
```
Comment (kuhar) [inline units/event_log_tests.cc]: Because this code is generic and will handle containers like `vector` or `string`, we should avoid needless copying.   We should take the value by a const reference.

File: units/event_log_tests.cc [hunk 23]
```diff
@@ -0,0 +1,183 @@
...
+  };
+};
+
+// Implements the Attribute base class for each specific `AttributeType`.
+// Valid typenames: {std::string, int64_t, std::vector<int64_t>}
+// Example:
+//    AttributeImpl<int64_t, AttributeType::Int64> timestamp("timestamp", 1234);
+template <typename T, AttributeType VT>
+struct AttributeImpl : Attribute {
+  T value;
```
Comment (kuhar) [inline units/event_log_tests.cc]: Similarly, could you make this a class and disallow modification to the underlying value? I would introduce a public const getter `const T& GetValue() const` and a non-const protected for derived classes `T& GetValue()`.

File: units/event_log_tests.cc [hunk 24]
```diff
@@ -0,0 +1,183 @@
...
+  template <typename T>
+  T *cast() {
+    return isa<T>() ? static_cast<T *>(this) : nullptr;
+  };
+};
+
+// Implements the Attribute base class for each specific `AttributeType`.
+// Valid typenames: {std::string, int64_t, std::vector<int64_t>}
+// Example:
+//    AttributeImpl<int64_t, AttributeType::Int64> timestamp("timestamp", 1234);
```
Comment (kuhar) [inline units/event_log_tests.cc]: Thanks, these examples look very useful. Could you surround them with code block, similar to `AttributeType` above?

File: units/event_log_tests.cc [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+    AttributeImpl<std::vector<int64_t>, AttributeType::VectorInt64>;
+using StringAttr = AttributeImpl<std::string, AttributeType::String>;
+using Int64Attr = AttributeImpl<int64_t, AttributeType::Int64>;
+
+// Event represents the base struct for a loggable event. It contains the
+// event's name and the level of importance. The derived structs must define and
+// initialize their own set of attributes.
+struct Event {
+  const char *name;
+  LogLevel log_level;
```
Comment (kuhar) [inline units/event_log_tests.cc]: Same here: class + deliberate access specifiers

File: units/event_log_tests.cc [hunk 26]
```diff
@@ -0,0 +1,183 @@
...
+};
+
+// An event for the CreateShaderModule function.
+// Example:
+//    const int64_t create_time_ns = create_end - create_start;
+//    Int64Attr timestamp("timestamp", 1234);
+//    Int64Attr shader_hash("hash", res.shader_hash);
+//    std::vector<Attribute *> attributes = {&shader_hash, &timestamp};
+//    CreateShaderModuleEvent("create_shader_module_ns", attributes,
+//    LogLevel::High);
```
Comment (kuhar) [inline units/event_log_tests.cc:107]: Could you surround this example with a code block?
````
```c++
your code
```
````

File: units/event_log_tests.cc [hunk 27]
```diff
@@ -0,0 +1,183 @@
...
+struct CreateShaderModuleEvent : Event {
+  std::vector<Attribute *> attributes;
+
+  CreateShaderModuleEvent(const char *_name,
+                          std::vector<Attribute *> &_attributes,
```
Comment (kuhar) [inline units/event_log_tests.cc]: I think it would be better to take make this event own all attributes. This would imply having a field per each attribute and setting them in the constructor.

File: units/event_log_tests.cc [hunk 28]
```diff
@@ -0,0 +1,183 @@
...
+  const std::vector<Attribute *> &GetAttributes() override {
+    return attributes;
+  }
+  size_t GetNumAttributes() override { return attributes.size(); }
+};
+
+// Takes an `Event` instance as an input and generates a csv string containing
+// `event`'s name and attribute values.
+std::string EventToCSVString(Event &event) {
+  const std::vector<Attribute *> &attributes = event.GetAttributes();
```
Comment (kuhar) [inline units/event_log_tests.cc]: Could you move all the `ValueToCSVString` just before this function? Keeping related code together will make it easier to understand.

File: units/event_log_tests.cc [hunk 29]
```diff
@@ -0,0 +1,183 @@
...
+
+  std::string csv_str = event.GetEventName();
+  csv_str += ",";
+  size_t e;
+  for (size_t i = 0, e = attributes.size(); i != e; ++i) {
+    switch (attributes[i]->value_type) {
+      case AttributeType::Int64: {
+        csv_str +=
+            ValueToCSVString(attributes[i]->cast<Int64Attr>()->GetValue());
```
Comment (kuhar) [inline units/event_log_tests.cc]: Use `std::stringstream` to build this string

File: units/event_log_tests.cc [hunk 30]
```diff
@@ -0,0 +1,183 @@
...
+        break;
+      }
+    }
+    if (i + 1 != e) csv_str += ",";
+  }
+  return csv_str;
+}
+
+TEST(Event, AttributeCreation) {
+  Int64Attr timestamp("timestamp", 1601314733266018318);
```
Comment (kuhar) [inline units/event_log_tests.cc]: In this basic test we could check that the value and name can be retrieved and that they match what you passed to create each of the attributes

File: units/event_log_tests.cc [hunk 31]
```diff
@@ -0,0 +1,183 @@
...
+
+TEST(Event, EventCreation) {
+  Int64Attr timestamp("timestamp", 1601314733266018318);
+  StringAttr state("state", "1");
+  VectorInt64Attr pipeline("pipeline",
+                           {0x67d6fd0aaa78a6d8, 0x67d390249c2f20ce});
+  std::vector<Attribute *> compile_attrs = {&timestamp, &state, &pipeline};
+  CreateShaderModuleEvent compile_event("compile_time", compile_attrs,
+                                        LogLevel::Low);
+  EXPECT_EQ(compile_event.GetNumAttributes(), 3);
```
Comment (kuhar) [inline units/event_log_tests.cc:229]: Use `ASSERT_EQ` instead. The difference is that `ASSERT_*` checks abort execution of the current test on failure, while `CHECK_*` continue executing. When the size is not as expected, we shouldn't try to access vector elements that don't exist.

File: units/event_log_tests.cc (hunk 27 above)
Comment (miladHakimi) [reply units/event_log_tests.cc]: Yeah. I wanted to do this in the next steps.

File: units/event_log_tests.cc [hunk 32]
```diff
@@ -0,0 +1,213 @@
...
+
+  // Safely casts the Attribute to the given class (T). Valid typenames should
+  // be derived from Attribute and have an ValueType `id_` field.
+  template <class T>
+  const T *cast() {
+    return isa<T>() ? static_cast<T *>(this) : nullptr;
+  };
+
+  const char *GetName() const { return name_; }
+  ValueType GetValueType() const { return value_type_; }
```
Comment (kuhar) [inline units/event_log_tests.cc:41]: Use `const ValueType& GetValueType() const` to avoid unnecessary copies

File: units/event_log_tests.cc [hunk 33]
```diff
@@ -0,0 +1,213 @@
...
+  return std::to_string(value);
+}
+
+std::string ValueToCSVString(const std::vector<int64_t> &values) {
+  std::stringstream csv_string;
+  csv_string << "\"[";
+  size_t e = values.size();
+  for (size_t i = 0; i != e; ++i) {
+    const char *delimiter = i < e - 1 ? "," : "";
+    csv_string << std::to_string(values[i]) + delimiter;
```
Comment (kuhar) [inline units/event_log_tests.cc]: No need to `to_string`, we can just append directly:
```c++
csv_string << values[i] << delimiter;
```

File: units/event_log_tests.cc [hunk 34]
```diff
@@ -0,0 +1,213 @@
...
+
+// Takes an `Event` instance as an input and generates a csv string containing
+// `event`'s name and attribute values.
+std::string EventToCSVString(Event &event) {
+  const std::vector<Attribute *> &attributes = event.GetAttributes();
+
+  std::stringstream csv_str;
+  csv_str << event.GetEventName();
+  csv_str << ",";
+  size_t e;
```
Comment (kuhar) [inline units/event_log_tests.cc]: No need to define `e` here, the `for` loop below already defines it

File: units/event_log_tests.cc [hunk 35]
```diff
@@ -0,0 +1,213 @@
...
+  return csv_str.str();
+}
+
+TEST(Event, AttributeCreation) {
+  const Int64Attr timestamp("timestamp", 1601314733266018318);
+  const StringAttr state("state", "1");
+  const VectorInt64Attr pipeline("pipeline",
+                                 {0x67d6fd0aaa78a6d8, 0x67d390249c2f20ce});
+  EXPECT_EQ(timestamp.GetName(), "timestamp");
+  EXPECT_EQ(timestamp.GetValue(), 1601314733266018318);
```
Comment (kuhar) [inline units/event_log_tests.cc]: It'd be better to put those long constants in local variables, so that it's immediately obvious when things should be equal, e.g.:
```c++
const int64_t timestamp_val = 160...;
const Int64Attr timestamp("timestamp", timestamp_val);
...
CHEKC_EQ(timestemp.GetValue(), timestamp_val);
```

File: units/event_log_tests.cc [hunk 36]
```diff
@@ -0,0 +1,213 @@
...
+  ASSERT_EQ(pipeline.GetValue().size(), 2);
+  EXPECT_EQ(pipeline.GetValue()[0], 0x67d6fd0aaa78a6d8);
+  EXPECT_EQ(pipeline.GetValue()[1], 0x67d390249c2f20ce);
+}
+
+TEST(Event, CreateShaderModuleEventCreation) {
+  CreateShaderModuleEvent compile_event("compile_time", 1601314733266018318,
+                                        0x67d6fd0aaa78a6d8, 926318,
+                                        LogLevel::Low);
+  ASSERT_EQ(compile_event.GetNumAttributes(), 3);
```
Comment (kuhar) [inline units/event_log_tests.cc]: Here `EXPECT_EQ` is fine because there is no other check that would otherwise fail.

**Review by kuhar — COMMENTED** [review]
The PR looks very nice now, just a few minor things

File: units/event_log_tests.cc (hunk 32 above)
Comment (miladHakimi) [reply units/event_log_tests.cc:41]: Since the ValueType's size was 4 bytes, I assumed the copy would be cheap. Sure, I'll change it.

Comment (kuhar) [reply units/event_log_tests.cc:41]: Ohhh, I got it wrong, I thought that this returns the underlying value. Please disregard then.

File: units/event_log_tests.cc [hunk 37]
```diff
@@ -0,0 +1,275 @@
...
+
+  // Safely casts the Attribute to the given class (T). Valid typenames should
+  // be derived from Attribute and have an ValueType `id_` field.
+  template <class T>
+  const T *cast() {
+    return isa<T>() ? static_cast<T *>(this) : nullptr;
+  };
+
+  const char *GetName() const { return name_; }
+  const ValueType &GetValueType() const { return value_type_; }
```
Comment (kuhar) [inline units/event_log_tests.cc]: Could you undo this change? I thought that this returned the underlying value, not the enum. Sorry for the confusion!

**Review by kuhar — APPROVED** [review]
Awesome, LGTM
